src/Futhark/IR/SOACS/SOAC.hs

```diff
@@ -91,10 +91,16 @@ data SOAC lore
     --
     -- The lambda body returns the output in this manner:
     --
-    --     [index_0, index_1, ..., index_n, value_0, value_1, ..., value_n]
+    --     [index_0, index_1, ..., index_n, value_0, value_1, ..., value_m]
     --
     -- This must be consistent along all Scatter-related optimisations.
-    Scatter SubExp (Lambda lore) [VName] [(SubExp, Int, VName)]
+    --
+    -- Scatters can be multi-dimensional, so the number of index-values need not
+    -- necessarily match the number of values. Instead, the number of
+    -- index-values should match the total number of indexes in the original
+    -- index and value arrays. The number of values should match the length of
+    -- the original index and value array.
```

Comment:
References to anything "original" is misleading.  The number of indexes must match the sum of the ranks of the shapes in the destination array list.